### ISSUE TITLE
guile heap blow-up on file write

### ISSUE THREAD

**Issue by linas** [MOZI-AI/annotation-scheme#111]
Sorry to pester you with so many issues, but .. I can't keep track of them all without writing them down.
I ran the following:
```
(gene-pathway-annotation gene-list  "my-path-results"
      #:pathway "reactome smpdb"
      #:namespace "biological_process molecular_function cellular_component"
      #:parents 0)
```
It took 60 hours to compute (2.5 days) During the entire time, the guile heap was at 150MBytes -- I know because I printed the heap size after each gene. Then it went to write the results file: this took 5 minutes to write, and the guile heap exploded to 16GB during those five minutes. 

The file write itself wrote 743843 results (this was the length of the "results" list)  The results file `xgene-pathway.scm` was 3974636585 bytes long (that's 4GBytes in size).

I conclude that there's something pathological about the file-writes; and perhaps things would go better if you wrote incrementally to a pipe, and then piped that pipe to a file.

**Comment by linas** [comment]
At this time, I'm seeing a blow-up here:  
https://github.com/MOZI-AI/annotation-scheme/blob/82c2bb4d76b1faf752b878a88283b096fa2d948f/annotation/main.scm#L120-L121

and it's the `(format #f "~a" result)` itself that is blowing up -- it requires a 10GByte string for the test case in issue #144 

As discussed elsewhere, the correct solution is probably to have callbacks: every time there is data to report, then call that callback and report it (and then the callback writes to file).  Do that, instead of using the giant-sized `append` or `concatenate` or `append-map` seen in the code.

**Comment by Habush** [comment]
> As discussed elsewhere, the correct solution is probably to have callbacks

I agree. But instead of callbacks I am thinking of using producer/consumer threads and using channels for communications, just like in [golang](https://golangbot.com/channels/). 
I know that such concept has been implemented in guile by Andy Wingo but haven't really used it. You can find the guile library [here](https://github.com/wingo/fibers)

**Comment by linas** [comment]
> producer/consumer threads

Yes! This seems quite reasonable. But you still need a callback, to let the consumer thread know that something is ready. (or to publish the results). The point of callbacks is you can use them for anything at all -- file-writes, database writes, producer/consumer threads, or some other queueing mechanism.  It's modular: if later on you decide to change how it's done, then there's only one small place to make the changes, instead of touching several dozen spots in the code.

**Comment by rekado** [comment]
Why not write to a port?

**Comment by rekado** [comment]
I wrote back when I rewrote the parser that it's still inefficient because there's a needless conversion from an s-expression to a string, and from a string back to an s-expression.  But we can simply read from a port instead.

**Comment by Habush** [comment]
> Why not write to a port?

Can we implement a producer/consumer b/n two threads just using ports? If so that will be a simpler solution. Can you please expand on the using a port solution? thanks.

**Comment by rekado** [comment]
My concern is with these lines:

https://github.com/MOZI-AI/annotation-scheme/blob/82c2bb4d76b1faf752b878a88283b096fa2d948f/annotation/main.scm#L120-L121

and

https://github.com/MOZI-AI/annotation-scheme/blob/master/annotation/parser.scm#L165

In `main.scm` there's the deadly string conversion.  And in `parser.scm` there's `with-input-from-string`.  Instead of using `with-input-from-string` on `port` (which is a string here, not a port), `atomese-parser` should be given an actual port, or even just the s-expression in `result`.

I'd first cut out that conversion before going with something more complicated.  (In that case I'd go with fibers.)

**Comment by Habush** [comment]
One thing I found out while trying to fix this issue is the parser, specifically the `expr->graph` function works only when the input is quoted. It doesn't work with s-expression that is not quoted, hence, it won't work with the output of the annotation functions. To this end,  I tried to quote the input like this:

```
(define* (atomese-parser expr)
  (set! *nodes* '())
  (set! *edges* '())
  (set! *atoms* '())
  (set! *annotation* "")
  (set! *prev-annotation* "")
  (for-each (lambda (e) (atomese->graph `,e)) expr)
  (make-graph *nodes* *edges*))
```

But this one has no effect. @rekado do you have any suggesstions?

**Comment by linas** [comment]
For readibility, just saying `(quote expr)` might be easier?

**Comment by linas** [comment]
So `(for-each (lambda (e) (atomese->graph (quote e))) expr)`

**Comment by Habush** [comment]
If I do `(quote e)` it will pass `e` to the parser not the atomese result

**Comment by rekado** [comment]
The backtick is syntax for quasiquote, the comma is unquote.  Doing `(quasiquote (unquote e))` is equivalent to just `e`.

`(quote e)` won't give you what you want, because that's just the symbol `e`.

The `atomese-parser` does indeed expect an S-expression.  Previously, we did get an S-expression as an argument here.  It was obtained through this odd conversion from data to string to S-expression (with `read` on the string).

What does the original expression look like?  If you don't have an S-expression to start with just match on the contents of the expression you do have.  It seems to me that this would simplify the code even more.

Remember that I only wrote the current implementation to incrementally simplify what came before.  It's not impossible, I think, to simplify this even more, perhaps even to a point where none of this is needed at all.

But to understand if that's possible we need to know: what does the input data look like and what do you want it to be converted to?  Once you can express this in a (possibly failing) test we could take a stab at finding a simpler solution to the actual needs.

Without tests that encode the requirements we're really just poking at the code without much purpose :)

**Comment by Habush** [comment]
> What does the original expression look like? If you don't have an S-expression to start with just match on the contents of the expression you do have. 

The original expression is Atomese. Something like `(ListLink (ConceptNode "A"))`

**Comment by Habush** [comment]
@linas how can we pass Atomese expressions like s-expression?

**Comment by linas** [comment]
> how can we pass Atomese expressions like s-expression?

I don't think you need to convert Atomese to s-exprs, do what @rekado said above:  https://github.com/MOZI-AI/annotation-scheme/issues/111#issuecomment-601408253

**Comment by Habush** [comment]
I made a fix for the above in #195. Please check it out.

**Comment by Habush** [comment]
Closing this.
